airflow/api_ui/security.py

```diff
@@ -0,0 +1,106 @@
+from airflow.auth.managers.models.resource_details import DagAccessEntity, DagDetails, DatasetDetails
+from airflow.providers.fab.auth_manager.api.auth.backend.basic_auth import auth_current_user
+from airflow.providers.fab.auth_manager.models import User
+from airflow.www.extensions.init_auth_manager import get_auth_manager
+
+security = HTTPBasic()
+
+
+def method(request: Request) -> ResourceMethod:
+    return cast(ResourceMethod, request.method)
+
+
+def check_authentication(
+    credentials: Annotated[HTTPBasicCredentials, Depends(security)],
+) -> User | None:
+    """Check that the request has valid authorization information."""
+    # TODO:
+    #    - Handle other auth backends
+    #    - Handle AUTH_ROLE_PUBLIC
+    user = auth_current_user(credentials)
+    if user is not None:
+        return user
+
+    # since this handler only checks authentication, not authorization,
+    # we should always return 401
+    raise HTTPException(401, headers={"WWW-Authenticate": "Basic"})
+
+
+def _requires_access(
+    *,
+    is_authorized_callback: Callable[[], bool],
+) -> None:
+    if not is_authorized_callback():
+        raise HTTPException(403, "Forbidden")
+
+
+def requires_access_dataset(
+    method: Annotated[ResourceMethod, Depends(method)],
+    uri: str | None = None,
+    user: Annotated[BaseUser | None, Depends(check_authentication)] = None,
+) -> None:
+    _requires_access(
+        is_authorized_callback=lambda: get_auth_manager().is_authorized_dataset(
+            user=user,
+            method=method,
+            details=DatasetDetails(uri=uri),
+        )
+    )
+
+
+def requires_access_dag(access_entity: DagAccessEntity | None = None) -> Callable:
+    def inner(
+        method: Annotated[ResourceMethod, Depends(method)],
+        dag_id: str | None = None,
+        user: Annotated[BaseUser | None, Depends(check_authentication)] = None,
+    ) -> None:
+        def callback():
+            access = get_auth_manager().is_authorized_dag(
+                method=method, access_entity=access_entity, details=DagDetails(id=dag_id), user=user
+            )
+
+            # ``access`` means here:
+            # - if a DAG id is provided (``dag_id`` not None): is the user authorized to access this DAG
+            # - if no DAG id is provided: is the user authorized to access all DAGs
+            if dag_id or access or access_entity:
+                return access
+
+            # No DAG id is provided, the user is not authorized to access all DAGs and authorization is done
+            # on DAG level
+            # If method is "GET", return whether the user has read access to any DAGs
+            # If method is "PUT", return whether the user has edit access to any DAGs
+            return (method == "GET" and any(get_auth_manager().get_permitted_dag_ids(methods=["GET"]))) or (
+                method == "PUT" and any(get_auth_manager().get_permitted_dag_ids(methods=["PUT"]))
+            )
```

Comment:
I remember when I moved this code I hated it and still hate it. I think we should probably simplify it with Airflow 3. We might want to return `True` in that case? That case is, the user is trying to access a list of DAGs and the user does not have permissions to see all DAGs. Here we are checking if the user has access to at least one, if not we return `AccessDenied`. To me, we should just return `True`, the function that retrieves the DAGs already filter in the DAGs that the user has access to. The only change would be, if a user does not has permissions to see DAGs, instead of having an access denied error, they would see an empty list of DAGs.

We dont need to solve it in this PR though.

Reply:
All the "list resources" scenarios (list DAGs, list connections, list variables ...) is not properly handled to me. Example: when listing connections, we are checking if the user has "GET" permissions on all connections. But the user might have "GET" permissions on one connection only, though in that case they would still have access denied when listing the connections. This is because in FAB you cannot give permissions on specific resources (besides DAGs) and we inherited from that. But we might want to fix that. Should we introduce a "LIST" permissions in auth managers?

Reply:
I understand your point and this makes sense. I do not have any strong opinion on that because I feel like both approach can be justified depending on how we want to build our permission model.

If a user request ressources 1, 2 and 3. If he is missing permissions on one of those ressources I can understand that the request is denied because he is basically forbidden from making that request.

Failing silently could have some weird user implications, for instance we return only the thing that you actually have access to i.e 1 and 2. Then the request can still be successful, but only contains partially the information that the user need. You have no other way than checking the entire content of the request to see if you are missing some informations or not.

Reply:
But I guess this is an interesting discussion for later. I feel like this is out of scope for this particular PR, I don't want to make adaptations to the permissions yet.